cmd/tether/attach_test.go

```diff
@@ -0,0 +1,613 @@
+}
+func (c *TestAttachTTYConfig) LoadConfig() (*metadata.ExecutorConfig, error) {
+	config := metadata.ExecutorConfig{}
+
+	config.ID = "attach"
+	config.Name = "tether_test_executor"
+	config.Sessions = map[string]metadata.SessionConfig{
+		"attach": metadata.SessionConfig{
+			Common: metadata.Common{
+				ID:   "attach",
+				Name: "tether_test_session",
+			},
+			Tty:    true,
+			Attach: true,
+			Cmd: metadata.Cmd{
+				Path: "/usr/bin/tee",
+				// grep, matching everything, reading from stdin
+				Args: []string{"/usr/bin/tee", pathPrefix + "/tee.out"},
+				Env:  []string{},
+				Dir:  "/",
+			},
+		},
+	}
+
+	// generate a host key for the tether
+	privateKey, err := rsa.GenerateKey(rand.Reader, 2014)
+	if err != nil {
+		return nil, err
+	}
+
+	privateKeyDer := x509.MarshalPKCS1PrivateKey(privateKey)
+	privateKeyBlock := pem.Block{
+		Type:    "RSA PRIVATE KEY",
+		Headers: nil,
+		Bytes:   privateKeyDer,
+	}
+
+	config.Key = pem.EncodeToMemory(&privateKeyBlock)
+
+	return &config, nil
+}
+
+func TestAttachTTY(t *testing.T) {
+	t.Skip("not sure how to test TTY yet")
+
+	// supply custom attach server so we can inspect its state
+	testServer := &testAttachServer{
+		updated: make(chan bool, 10),
+	}
+	server = testServer
+
+	testSetup(t)
+	defer testTeardown(t)
+
+	// if there's no session command with guaranteed exit then tether needs to run in the background
+	cfg := &TestAttachTTYConfig{}
+	testConfig, err := cfg.LoadConfig()
+	if err != nil {
+		t.Error(err)
+		return
+	}
+
+	go func() {
+		err := run(cfg)
+		if err != nil {
+			t.Error(err)
+		}
+	}()
+
+	// wait for updates to occur
+	<-testServer.updated
+
+	if !testServer.enabled {
+		t.Error("attach server was not enabled")
+		return
+	}
+
+	// create client on the mock pipe
+	conn, err := clientBackchannel(context.Background())
+	if err != nil {
+		t.Error(err)
+		return
+	}
+
+	cconfig := &ssh.ClientConfig{
+		User: "daemon",
+		HostKeyCallback: func(hostname string, remote net.Addr, key ssh.PublicKey) error {
+			return nil
+		},
+	}
+
+	// create the SSH client
+	sConn, chans, reqs, err := ssh.NewClientConn(conn, "notappliable", cconfig)
+	if err != nil {
+		t.Error(err)
+		return
+	}
+	defer sConn.Close()
+	client := ssh.NewClient(sConn, chans, reqs)
+
+	session, err := SSHAttach(client, testConfig.ID)
+	if err != nil {
+		t.Error(err)
+		return
+	}
+
+	stdout := session.Stdout()
+
+	// FIXME: this is line buffered - how do I disable that so we don't have odd hangs to diagnose
+	// when the trailing \n is missed
+	testBytes := []byte("hello world!\n")
+	// read from session into buffer
+	buf := &bytes.Buffer{}
+
+	var wg sync.WaitGroup
+	go func() { wg.Add(1); io.CopyN(buf, stdout, int64(len(testBytes))); wg.Done() }()
+
+	// write something to echo
+	log.Debug("sending test data")
+	session.Stdin().Write(testBytes)
+	log.Debug("sent test data")
+
+	// wait for the close to propogate
+	wg.Wait()
+	session.Stdin().Close()
+
+	if !bytes.Equal(buf.Bytes(), testBytes) {
```

Comment:
The assert package has helpers for this kind of stuff.  `if !assert.Equal(t, expected, actual) { return }`.  And it'll print a unified diff of the objects' differences if they aren't equal.